File: internal/api/jdwp/classloaderreference.go
package jdwp

import (
	"fmt"

	"cli-debugger/pkg/errors"
)

// ClassLoaderReference Command Set Implementation
// ClassLoaderReference command constants are already defined in protocol.go

// VisibleClasses Get visible classes
func (c *Client) VisibleClasses(classLoaderID string) ([]*ClassInfo, error) {
	data := encodeID(classLoaderID, c.idsizes.ObjectIDSize)

	packet := createCommandPacketWithData(classLoaderCommandSet, classLoaderCommandVisibleClasses, data)
	if err := c.sendPacket(packet); err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get visible classes")
	}

	reply, err := c.readReply()
	if err != nil {
		return nil, errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get visible classes")
	}

	if reply.ErrorCode != 0 {
		return nil, errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get visible classes failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	classCount := reader.readInt()

	classes := make([]*ClassInfo, 0, classCount)
	for i := 0; i < classCount; i++ {
		tag := reader.readByte()
		refID := reader.readID(c.idsizes.ReferenceTypeIDSize)
		status := reader.readInt()

		classes = append(classes, &ClassInfo{
			Tag:    tag,
			RefID:  refID,
			Status: status,
		})
	}

	return classes, nil
}

File: src/protocol/jdwp/class-loader-reference.ts
/**
 * ClassLoaderReference Command Set Implementation
 */

import {
  CommandSet,
  ClassLoaderReferenceCommand,
  type IDSizes,
} from "./protocol.js";
import {
  createCommandPacketWithData,
  encodeID,
} from "./codec.js";
import { PacketReader } from "./reader.js";
import { APIError, ErrorType, ErrorCodes } from "../errors.js";

/**
 * JDWP Client interface for command execution
 */
export interface JDWPCommandExecutor {
  sendPacket(packet: Buffer): Promise<void>;
  readReply(): Promise<{ errorCode: number; message: string; data: Buffer }>;
  idSizes: IDSizes;
}

/**
 * Get visible classes for a class loader
 */
export async function getVisibleClasses(
  executor: JDWPCommandExecutor,
  classLoaderID: string,
): Promise<{
  classes: Array<{
    refTypeID: string;
    typeTag: number;
    status: number;
  }>;
}> {
  const data = encodeID(classLoaderID, executor.idSizes.objectIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.ClassLoaderReference,
    ClassLoaderReferenceCommand.VisibleClasses,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get visible classes failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  const classesCount = reader.readInt();

  const classes: Array<{
    refTypeID: string;
    typeTag: number;
    status: number;
  }> = [];

  for (let i = 0; i < classesCount; i++) {
    const typeTag = reader.readByte();
    const refTypeID = reader.readID(executor.idSizes.referenceTypeIDSize);
    const status = reader.readInt();

    classes.push({
      refTypeID,
      typeTag,
      status,
    });
  }

  return {
    classes,
  };
}
